### PR TITLE
Adding Type Hints to Captum: saliency.py/test_salientcy.py

### PR DESCRIPTION
**Pull request by icpm**
This PR adds Type Hints to Captum: saliency.py and its test test_saliency.py.

### CODE REVIEW

File: tests/attr/test_saliency.py [hunk 1]
```diff
@@ -109,11 +112,12 @@ def _saliency_classification_assert(self, nt_type="vanilla"):
 
             output = model(input)[:, target]
             output.backward()
-            expected = torch.abs(input.grad)
-            self.assertEqual(
-                expected.detach().numpy().tolist(),
-                attributions.detach().numpy().tolist(),
-            )
+            if input.grad:
```
Comment (icpm) [inline tests/attr/test_saliency.py]: `input.grad` is typed as Optional[Tensor], and is shouldn't be None by any chance. Thus I add a if check here to make mypy happy

File: tests/attr/test_saliency.py [hunk 2]
```diff
@@ -76,27 +85,21 @@ def _saliency_base_assert(
...
-        if isinstance(attributions, tuple):
-            for input, attribution, expected_attr in zip(
-                inputs, attributions, expected
-            ):
-                if nt_type == "vanilla":
-                    self._assert_attribution(attribution, expected_attr)
-                self.assertEqual(input.shape, attribution.shape)
-        else:
+
+        for input, attribution, expected_attr in zip(inputs, attributions, expected):
```
Comment (icpm) [inline tests/attr/test_saliency.py]: I found that the if/else check is redundant in this case -- no matter the `attributions` is a tuple or not, we can use one for loop to iterate over. In addition, this refactor makes mypy happy with types of attributions/inputs

File: captum/attr/_core/saliency.py [hunk 3]
```diff
@@ -17,7 +21,15 @@ def __init__(self, forward_func):
...
         GradientAttribution.__init__(self, forward_func)
 
-    def attribute(self, inputs, target=None, abs=True, additional_forward_args=None):
+    def attribute(
+        self,
+        inputs: TensorOrTupleOfTensors,
+        target: Optional[
+            Union[int, Tuple[int, ...], Tensor, List[Tuple[int, ...]]]
+        ] = None,
+        abs: Optional[bool] = True,
```
[the remark this answers is not in the document]
Comment (icpm) [reply captum/attr/_core/saliency.py]: sure

File: tests/attr/test_saliency.py [hunk 4]
```diff
@@ -1,67 +1,81 @@
...
+from typing import Any, Tuple, cast
 
-from captum.attr._core.saliency import Saliency
+import torch
 from captum.attr._core.noise_tunnel import NoiseTunnel
+from captum.attr._core.saliency import Saliency
 from captum.attr._utils.gradient import compute_gradients
+from captum.attr._utils.typing import TensorOrTupleOfTensors
+from torch import Tensor
+from torch.nn import Module
```
[the remark this answers is not in the document]
Comment (icpm) [reply tests/attr/test_saliency.py]: absolutely

File: tests/attr/test_saliency.py [hunk 5]
```diff
@@ -1,67 +1,81 @@
...
+]:
     model = BasicModel5_MultiArgs()
     additional_forward_args = ([2, 3], 1)
     inputs = (
         torch.tensor([[1.5, 2.0, 34.3], [3.4, 1.2, 2.0]], requires_grad=True),
         torch.tensor([[3.0, 3.5, 23.2], [2.3, 1.2, 0.3]], requires_grad=True),
     )
-    grads = compute_gradients(
-        model, inputs, additional_forward_args=additional_forward_args
+    grads = cast(
```
[the remark this answers is not in the document]
Comment (icpm) [reply tests/attr/test_saliency.py]: I type hint the compute_gradients function instead using cast here

File: captum/attr/_utils/gradient.py [hunk 6]
```diff
@@ -62,8 +67,11 @@ def undo_gradient_requirements(inputs, grad_required):
 
 
 def compute_gradients(
-    forward_fn, inputs, target_ind=None, additional_forward_args=None
-):
+    forward_fn: Module,
+    inputs: TensorOrTupleOfTensors,
+    target_ind: Optional[TensorOrTupleOfTensors] = None,
```
Comment (vivekmig) [inline captum/attr/_utils/gradient.py]: This type hint should match target, e.g. Optional[Union[int, Tuple[int, ...], Tensor, List[Tuple[int, ...]]]]
